### PR TITLE
add GPU support to runtime/LibTorch

### PR DESCRIPTION
**Pull request by veelion**
Support to build C++  runtime/LibTorch   with libtorch-cu11.3, to make websocket_server_main can run on GPU.

### CODE REVIEW

**Comment by robin1001** [comment]
Any RTF(real time factor) benchmark result?

File: runtime/core/cmake/libtorch.cmake [hunk 1]
```diff
@@ -54,4 +64,7 @@ if(TORCH)
     )
   endif()
   add_definitions(-DUSE_TORCH)
+  if(GPU)
+    add_definitions(-DUSE_GPU)
```
Comment (pengzhendong) [inline runtime/core/cmake/libtorch.cmake:68]: I think set the cuda version here is better.
BTW, maybe you should print some message when `USE_GPU` in Windows and MacOS.

**Comment by veelion** [comment]
Test ENV: (docker container)

- 15 CPU core of AMD EPYC 7543 32-Core Processor
- NVIDIA-SMI 510.60.02    Driver Version: 510.60.02    CUDA Version: 11.3
- GPU: A5000*1
- test data: 3000 wavs from LeaderBoard test data, total duration is 16455s
- runtime: websocket_server_main

| client concurrency | RTF | GPU usage |
| --- | --- | --- |
| 200 | 0.00529 | 0% - 52% |
| 300 | 0.00546 | 0% - 91% |
| 100 | 0.00478 | 1% - 54% |
| 200 | 0.00492 | 0% - 52% |
| 300 | 0.00492 | 0% - 91% |

File: runtime/core/cmake/libtorch.cmake (hunk 1 above)
Comment (veelion) [reply runtime/core/cmake/libtorch.cmake:68]: like cmake -DUSE_GPU=11.3 ？
just  print 'not support yet', is it ok?

**Comment by robin1001** [comment]
What's the chunk size in your testing?

**Comment by veelion** [comment]
> 

chunk size is -1. updated in the test result above.

**Comment by robin1001** [comment]
> > 
> 
> chunk size is -1. updated in the test result above.

-1 is for non-streaming decoding, and non-streaming shows better RTF than streaming. I wonder if the RTF drop a lot in a streaming testing. You can try chunk_size 16, which is for streaming testing.

File: runtime/core/decoder/torch_asr_model.cc [hunk 2]
```diff
@@ -41,6 +42,17 @@ void TorchAsrModel::Read(const std::string& model_path) {
   model_ = std::make_shared<TorchModule>(std::move(model));
   torch::NoGradGuard no_grad;
   model_->eval();
+#ifdef USE_GPU
```
[the remark this answers is not in the document]
Comment (veelion) [reply runtime/core/decoder/torch_asr_model.cc:45]: USE_GPU macro tell compiler only build GPU codes, so is only for GPU inference. (one build for on running scenario, with only one libtorch)

If we want to choose cpu or gpu after building, there are 2 ways, I guess, 

1. when start websocket_server_main, add an argument to tell the programme to use CPU or GPU. (one build for two running scenario, with two version of libtorch for cpu and gpu)
2. use `torch::cuda::is_available()` to choose CPU or GPU automatically. ( GPU first, CPU if no GPU)

I think one build for on scenario is the good choice for deploying it has smallest file size, no condition checking for GPU or CPU. Before we deploy the ASR service, we should make a decision to use CPU or GPU, according the amount of audio to be transcript, the concurrency of request.

If we USE_GPU while building, I think it is no need to add argu `device` to `Read()`, but `torch::jit::load(model_path, device)` is better than `model_.to(at::kCUDA)`, I will try it. I think it's better to keep python codes unchanged, only change C++ codes for C++ runtime.

**Comment by yuekaizhang** [comment]
Could you please also update https://github.com/wenet-e2e/wenet/blob/main/runtime/LibTorch/README.md? Maybe one section about how to build and run with gpu inference?

**Review by yuekaizhang — CHANGES_REQUESTED** [review]
Looks good to me. After updating the readme, I think this PR could be merged.

**Comment by veelion** [comment]
> 

streaming test got error, a guy in the Wechat group also got the same error.
running command:
```
./websocket_server_main --port 8301 --chunk_size 16 --model_path $model_dir/final.zip --unit_path $model_dir/words.txt 2>&1 | tee server.log

./websocket_client_main --hostname 127.0.0.1 --port 8301 --wav_path $wav_path --continuous_decoding
```
server log message:

```
I0715 10:56:42.633028 618355 websocket_server.cc:212] {"signal":"start","nbest":1,"continuous_decoding":true}
I0715 10:56:42.633153 618355 websocket_server.cc:43] Received speech start signal, start reading speech
I0715 10:56:42.635293 618355 websocket_server.cc:90] Received 8000 samples
I0715 10:56:43.135967 618355 websocket_server.cc:90] Received 8000 samples   
I0715 10:56:43.140075 618357 asr_decoder.cc:104] Required 67 get 67       
I0715 10:56:43.456400 618357 asr_decoder.cc:194] Partial CTC result 感觉这
I0715 10:56:43.456493 618357 websocket_server.cc:65] Partial result: [{"sentence":"感觉这"}]
I0715 10:56:43.636518 618355 websocket_server.cc:90] Received 8000 samples
I0715 10:56:43.640735 618357 asr_decoder.cc:104] Required 64 get 64
I0715 10:56:44.137161 618355 websocket_server.cc:90] Received 8000 samples 
I0715 10:56:44.637842 618355 websocket_server.cc:90] Received 8000 samples   
I0715 10:56:45.138478 618355 websocket_server.cc:90] Received 8000 samples
I0715 10:56:45.639073 618355 websocket_server.cc:90] Received 8000 samples
I0715 10:56:46.139631 618355 websocket_server.cc:90] Received 8000 samples
I0715 10:56:46.640404 618355 websocket_server.cc:90] Received 8000 samples
I0715 10:56:47.141006 618355 websocket_server.cc:90] Received 8000 samples
I0715 10:56:47.641297 618355 websocket_server.cc:90] Received 3121 samples
I0715 10:56:48.141386 618355 websocket_server.cc:212] {"signal":"end"}   
I0715 10:56:48.141461 618355 websocket_server.cc:57] Received speech end signal
I0715 10:56:48.141469 618355 websocket_server.cc:229] Read all pcm data, wait for decoding thread
E0715 10:56:48.592036 618357 websocket_server.cc:152] The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File "code/__torch__/wenet/transformer/asr_model.py", line 148, in fallback_function
    cnn_cache: Tensor=CONSTANTS.c0) -> Tuple[Tensor, Tensor, Tensor]:
    encoder = self.encoder                                                                                        
    _25 = (encoder).forward_chunk(xs, offset, required_cache_size, att_cache, cnn_cache, CONSTANTS.c1, )
           ~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    return _25
  def is_bidirectional_decoder(self: __torch__.wenet.transformer.asr_model.ASRModel) -> bool:
  File "code/__torch__/wenet/transformer/encoder.py", line 136, in forward_chunk
    else:
      _46 = cnn_cache
    _47 = (_0).forward(xs16, att_mask, pos_emb0, CONSTANTS.c1, _44, _46, )
           ~~~~~~~~~~~ <--- HERE
    xs17, _48, new_att_cache, new_cnn_cache, = _47
    _49 = torch.slice(torch.slice(new_att_cache), 1)
  File "code/__torch__/wenet/transformer/encoder_layer.py", line 53, in forward
      x3 = x2
    self_attn = self.self_attn
    _2 = (self_attn).forward(x3, x3, x3, mask, pos_emb, att_cache, )
          ~~~~~~~~~~~~~~~~~~ <--- HERE
    x_att, new_att_cache, = _2
    concat_after = self.concat_after
  File "code/__torch__/wenet/transformer/attention.py", line 46, in forward
    matrix_ac = torch.matmul(q_with_bias_u, torch.transpose(k0, -2, -1))
    matrix_bd = torch.matmul(q_with_bias_v, torch.transpose(p0, -2, -1))
    _3 = torch.add(matrix_ac, matrix_bd)
         ~~~~~~~~~ <--- HERE
    d_k0 = self.d_k
    scores = torch.div(_3, torch.sqrt(d_k0))
                                                                                                                  
Traceback of TorchScript, original code (most recent call last):                   
  File "/root/wenet/wenet/transformer/asr_model.py", line 614, in fallback_function                               
           
        """                                                                                                       
        return self.encoder.forward_chunk(xs, offset, required_cache_size,                                        
               ~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE                                                               
                                          att_cache, cnn_cache)              
  File "/root/wenet/wenet/transformer/encoder.py", line 249, in forward_chunk
            #   shape(att_cache[i:i + 1]) is (1, head, cache_t1, d_k * 2),
            #   shape(cnn_cache[i])       is (b=1, hidden-dim, cache_t2)                                          
            xs, _, new_att_cache, new_cnn_cache = layer(                                                          
                                                  ~~~~~ <--- HERE                                                 
                xs, att_mask, pos_emb,                                                                            
                att_cache=att_cache[i:i + 1] if elayers > 0 else att_cache,  
  File "/root/wenet/wenet/transformer/encoder_layer.py", line 234, in forward                                     
            x = self.norm_mha(x)
                                              
        x_att, new_att_cache = self.self_attn(         
                               ~~~~~~~~~~~~~~ <--- HERE
            x, x, x, mask, pos_emb, att_cache)
        if self.concat_after:                                                                                     
  File "/root/wenet/wenet/transformer/attention.py", line 309, in forward                                         
        # matrix_bd = self.rel_shift(matrix_bd)
     
        scores = (matrix_ac + matrix_bd) / math.sqrt(
                  ~~~~~~~~~~~~~~~~~~~~~ <--- HERE
            self.d_k)  # (batch, head, time1, time2)
     
RuntimeError: The size of tensor a (32) must match the size of tensor b (0) at non-singleton dimension 3
```

**Comment by veelion** [comment]
> > 
> 
> streaming test got error, a guy in the Wechat group also got the same error. running command:
> 
> ```
> ./websocket_server_main --port 8301 --chunk_size 16 --model_path $model_dir/final.zip --unit_path $model_dir/words.txt 2>&1 | tee server.log
> 
> ./websocket_client_main --hostname 127.0.0.1 --port 8301 --wav_path $wav_path --continuous_decoding
> ```
> 
> server log message:
> 
> ```
> I0715 10:56:42.633028 618355 websocket_server.cc:212] {"signal":"start","nbest":1,"continuous_decoding":true}
> I0715 10:56:42.633153 618355 websocket_server.cc:43] Received speech start signal, start reading speech
> I0715 10:56:42.635293 618355 websocket_server.cc:90] Received 8000 samples
> I0715 10:56:43.135967 618355 websocket_server.cc:90] Received 8000 samples   
> I0715 10:56:43.140075 618357 asr_decoder.cc:104] Required 67 get 67       
> I0715 10:56:43.456400 618357 asr_decoder.cc:194] Partial CTC result 感觉这
> I0715 10:56:43.456493 618357 websocket_server.cc:65] Partial result: [{"sentence":"感觉这"}]
> I0715 10:56:43.636518 618355 websocket_server.cc:90] Received 8000 samples
> I0715 10:56:43.640735 618357 asr_decoder.cc:104] Required 64 get 64
> I0715 10:56:44.137161 618355 websocket_server.cc:90] Received 8000 samples 
> I0715 10:56:44.637842 618355 websocket_server.cc:90] Received 8000 samples   
> I0715 10:56:45.138478 618355 websocket_server.cc:90] Received 8000 samples
> I0715 10:56:45.639073 618355 websocket_server.cc:90] Received 8000 samples
> I0715 10:56:46.139631 618355 websocket_server.cc:90] Received 8000 samples
> I0715 10:56:46.640404 618355 websocket_server.cc:90] Received 8000 samples
> I0715 10:56:47.141006 618355 websocket_server.cc:90] Received 8000 samples
> I0715 10:56:47.641297 618355 websocket_server.cc:90] Received 3121 samples
> I0715 10:56:48.141386 618355 websocket_server.cc:212] {"signal":"end"}   
> I0715 10:56:48.141461 618355 websocket_server.cc:57] Received speech end signal
> I0715 10:56:48.141469 618355 websocket_server.cc:229] Read all pcm data, wait for decoding thread
> E0715 10:56:48.592036 618357 websocket_server.cc:152] The following operation failed in the TorchScript interpreter.
> Traceback of TorchScript, serialized code (most recent call last):
>   File "code/__torch__/wenet/transformer/asr_model.py", line 148, in fallback_function
>     cnn_cache: Tensor=CONSTANTS.c0) -> Tuple[Tensor, Tensor, Tensor]:
>     encoder = self.encoder                                                                                        
>     _25 = (encoder).forward_chunk(xs, offset, required_cache_size, att_cache, cnn_cache, CONSTANTS.c1, )
>            ~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
>     return _25
>   def is_bidirectional_decoder(self: __torch__.wenet.transformer.asr_model.ASRModel) -> bool:
>   File "code/__torch__/wenet/transformer/encoder.py", line 136, in forward_chunk
>     else:
>       _46 = cnn_cache
>     _47 = (_0).forward(xs16, att_mask, pos_emb0, CONSTANTS.c1, _44, _46, )
>            ~~~~~~~~~~~ <--- HERE
>     xs17, _48, new_att_cache, new_cnn_cache, = _47
>     _49 = torch.slice(torch.slice(new_att_cache), 1)
>   File "code/__torch__/wenet/transformer/encoder_layer.py", line 53, in forward
>       x3 = x2
>     self_attn = self.self_attn
>     _2 = (self_attn).forward(x3, x3, x3, mask, pos_emb, att_cache, )
>           ~~~~~~~~~~~~~~~~~~ <--- HERE
>     x_att, new_att_cache, = _2
>     concat_after = self.concat_after
>   File "code/__torch__/wenet/transformer/attention.py", line 46, in forward
>     matrix_ac = torch.matmul(q_with_bias_u, torch.transpose(k0, -2, -1))
>     matrix_bd = torch.matmul(q_with_bias_v, torch.transpose(p0, -2, -1))
>     _3 = torch.add(matrix_ac, matrix_bd)
>          ~~~~~~~~~ <--- HERE
>     d_k0 = self.d_k
>     scores = torch.div(_3, torch.sqrt(d_k0))
>                                                                                                                   
> Traceback of TorchScript, original code (most recent call last):                   
>   File "/root/wenet/wenet/transformer/asr_model.py", line 614, in fallback_function                               
>            
>         """                                                                                                       
>         return self.encoder.forward_chunk(xs, offset, required_cache_size,                                        
>                ~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE                                                               
>                                           att_cache, cnn_cache)              
>   File "/root/wenet/wenet/transformer/encoder.py", line 249, in forward_chunk
>             #   shape(att_cache[i:i + 1]) is (1, head, cache_t1, d_k * 2),
>             #   shape(cnn_cache[i])       is (b=1, hidden-dim, cache_t2)                                          
>             xs, _, new_att_cache, new_cnn_cache = layer(                                                          
>                                                   ~~~~~ <--- HERE                                                 
>                 xs, att_mask, pos_emb,                                                                            
>                 att_cache=att_cache[i:i + 1] if elayers > 0 else att_cache,  
>   File "/root/wenet/wenet/transformer/encoder_layer.py", line 234, in forward                                     
>             x = self.norm_mha(x)
>                                               
>         x_att, new_att_cache = self.self_attn(         
>                                ~~~~~~~~~~~~~~ <--- HERE
>             x, x, x, mask, pos_emb, att_cache)
>         if self.concat_after:                                                                                     
>   File "/root/wenet/wenet/transformer/attention.py", line 309, in forward                                         
>         # matrix_bd = self.rel_shift(matrix_bd)
>      
>         scores = (matrix_ac + matrix_bd) / math.sqrt(
>                   ~~~~~~~~~~~~~~~~~~~~~ <--- HERE
>             self.d_k)  # (batch, head, time1, time2)
>      
> RuntimeError: The size of tensor a (32) must match the size of tensor b (0) at non-singleton dimension 3
> ```

It's my mistake in changing c++ cods. Fixed and continue streaming testing.

**Comment by yuekaizhang** [comment]
I tested it with single wav, however, the program looks exit unnormally. I can't see the log from https://github.com/wenet-e2e/wenet/blob/main/runtime/core/bin/decoder_main.cc#L160-L162. Could you also verify it?
```bash
./build/bin/decoder_main \
 --chunk_size -1 \
 --wav_path $wav_path \
 --model_path $model_dir/final.zip \
 --unit_path $model_dir/units.txt 2>&1 | tee log.txt
```

**Comment by veelion** [comment]
> I tested it with single wav, however, the program has not been properly exited. I can't see the log from https://github.com/wenet-e2e/wenet/blob/main/runtime/core/bin/decoder_main.cc#L160-L162. Could you also verify it? [log.txt](https://github.com/wenet-e2e/wenet/files/9117669/log.txt)
> 
> ```shell
> ./build/bin/decoder_main \
>  --chunk_size -1 \
>  --wav_path $wav_path \
>  --model_path $model_dir/final.zip \
>  --unit_path $model_dir/units.txt 2>&1 | tee log.txt
> ```

I run a few times, and each runing can see the log msg.

**Comment by veelion** [comment]
Test-2: Test ENV: (docker container)

- 15 CPU core of AMD EPYC 7543 32-Core Processor
- NVIDIA-SMI 510.60.02 Driver Version: 510.60.02 CUDA Version: 11.3
- GPU: A5000*1
- test data: wavs from LeaderBoard test data, total duration is 16455s
- runtime: decoder_main

| audio duration | chunk_size=-1 | chunk_size=16 |
| --- | --- | --- |
| 5 | 0.08798 | 2.971 |
| 7 | 0.07208 | 2.874 |
|60| 0.02517 | 0.6659 |
| 45 (10 wavs) | 1.18 | 1.222 |
| 604 (100 wavs) | 0.2093 | 0.2081 |

**Comment by yuekaizhang** [comment]
Attach my test results, using one num_thread, repeatly decoding one audio until the decode time gets stable. 

| audio duration | GPU decode duration(chunk_size=-1) | GPU decode duration(chunk_size=16) | CPU decode duration(chunk_size=-1) | CPU decode duration(chunk_size=16) |
| --- | --- | --- | --- | --- |
| 4992ms | 112ms |  231 ms | 384ms| 553ms |

I think when decoding with chunk_size 16, data transfer time between GPU and CPU would be longer comparing with offline ASR. 

Also, later, if you or someone could support batch decode using GPU would be very welcome.

**Review by yuekaizhang — APPROVED** [review]

**Pull request merged**